### PR TITLE
[FLINK-8432] [filesystem-connector] Add support for openstack's swift filesystem

### PR DESCRIPTION
**Pull request by jelmerk**
## What is the purpose of the change

Add support for OpenStack's cloud storage solution without Hadoop dependencies

## Brief change log

- Added new module below flink-filesystems

## Verifying this change

This change added tests and can be verified as follows:

- Added integration tests for simple reading and writing and listing directories

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: don't know
  - The runtime per-record code paths (performance sensitive): don't know
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: don't know
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by jelmerk** [comment]
Hi @tillrohrmann Those files with a lot of changes  are unfortunately copied from hadoop, the readme file contains more information about how this works. I am not overly happy about the this approach. But its basically the exact same approach that was taken for the s3 connector https://github.com/apache/flink/tree/master/flink-filesystems/flink-s3-fs-hadoop

File: flink-filesystems/flink-openstack-fs-hadoop/src/test/java/org/apache/flink/fs/openstackhadoop/HadoopSwiftFileSystemITCase.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Integration tests for the Swift file system support.
+ */
+public class HadoopSwiftFileSystemITCase extends TestLogger {
```
Comment (tillrohrmann) [inline flink-filesystems/flink-openstack-fs-hadoop/src/test/java/org/apache/flink/fs/openstackhadoop/HadoopSwiftFileSystemITCase.java]: I assume yo run these tests locally on OpenStack, right?

**Comment by etiennecarriere** [comment]
We tested this PR with flink 1.4 and the swift offered by French Hoster OVH. 

I was fine for checkpoint .

File: flink-filesystems/flink-openstack-fs-hadoop/src/main/java/org/apache/flink/fs/openstackhadoop/SwiftFileSystemFactory.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+			org.apache.hadoop.conf.Configuration hadoopConfig = this.hadoopConfig;
+			if (hadoopConfig == null) {
+				if (flinkConfig != null) {
+					LOG.debug("Loading Hadoop configuration for swift native file system");
+					hadoopConfig = HadoopUtils.getHadoopConfiguration(flinkConfig);
+
+					// hadoop.tmp.dir needs to be defined because it is used as buffer directory
+					if (hadoopConfig.get("hadoop.tmp.dir") == null) {
+						String tmpDir = System.getProperty("java.io.tmpdir") + "/" + "hadoop-" + System.getProperty("user.name");
```
[the remark this answers is not in the document]
Comment (jelmerk) [reply flink-filesystems/flink-openstack-fs-hadoop/src/main/java/org/apache/flink/fs/openstackhadoop/SwiftFileSystemFactory.java]: Changed in 718ac46f45bc0b1a275f301255c52cec7ecdd209

File: flink-filesystems/flink-openstack-fs-hadoop/src/main/java/org/apache/flink/fs/openstackhadoop/SwiftFileSystemFactory.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+						hadoopConfig.set("hadoop.tmp.dir", tmpDir);
+					}
+
+					// add additional config entries from the Flink config to the Presto Hadoop config
+					for (String key : flinkConfig.keySet()) {
+						if (key.startsWith(CONFIG_PREFIX)) {
+							String value = flinkConfig.getString(key, null);
+							String newKey = "fs.swift." + key.substring(CONFIG_PREFIX.length());
+							hadoopConfig.set(newKey, flinkConfig.getString(key, null));
```
[the remark this answers is not in the document]
Comment (jelmerk) [reply flink-filesystems/flink-openstack-fs-hadoop/src/main/java/org/apache/flink/fs/openstackhadoop/SwiftFileSystemFactory.java]: Changed in 718ac46f45bc0b1a275f301255c52cec7ecdd209

File: flink-filesystems/flink-openstack-fs-hadoop/src/test/java/org/apache/flink/fs/openstackhadoop/HadoopSwiftFileSystemITCase.java (hunk 1 above)
Comment (jelmerk) [reply flink-filesystems/flink-openstack-fs-hadoop/src/test/java/org/apache/flink/fs/openstackhadoop/HadoopSwiftFileSystemITCase.java]: I did

**Comment by jelmerk** [comment]
All valid points, thanks for the review!

**Comment by tillrohrmann** [comment]
Great. Thanks a lot for addressing my comments so fast @jelmerk and @etiennecarriere. Thanks a lot for your contribution. Merging this PR.

**Pull request closed without merging**